react/features/recording/components/Recording/AbstractRecordButton.js

```diff
@@ -0,0 +1,127 @@
+import {
+    getLocalParticipant,
+    isLocalParticipantModerator
+} from '../../../base/participants';
+import {
+    AbstractButton,
+    type AbstractButtonProps
+} from '../../../base/toolbox';
+
+import { getActiveSession } from '../../functions';
+
+import StartRecordingDialog from './StartRecordingDialog';
+import StopRecordingDialog from './StopRecordingDialog';
+
+/**
+ * The type of the React {@code Component} props of
+ * {@link AbstractRecordButton}.
+ */
+export type Props = AbstractButtonProps & {
+
+    /**
+     * True if there is a running active recording, false otherwise.
+     */
+    _isRecordingRunning: boolean,
+
+    /**
+     * The redux {@code dispatch} function.
+     */
+    dispatch: Function,
+
+    /**
+     * The i18n translate function.
+     */
+    t: Function
+};
+
+/**
+ * An abstract implementation of a button for starting and stopping recording.
+ */
+export default class AbstractRecordButton<P: Props>
+    extends AbstractButton<P, *> {
+    accessibilityLabel = 'Recording';
+    label = 'dialog.startRecording';
+    toggledLabel = 'dialog.stopRecording';
+
+    /**
+     * Handles clicking / pressing the button.
+     *
+     * @override
+     * @protected
+     * @returns {void}
+     */
+    _handleClick() {
+        const { _isRecordingRunning, dispatch } = this.props;
+
+        sendAnalytics(createToolbarEvent(
+            'recording.button',
+            {
+                'is_recording': _isRecordingRunning,
+                type: JitsiRecordingConstants.mode.FILE
+            }));
+
+        dispatch(openDialog(
+            _isRecordingRunning ? StopRecordingDialog : StartRecordingDialog
+        ));
+    }
+
+    /**
+     * Helper function to be implemented by subclasses, which must return a
+     * boolean value indicating if this button is disabled or not.
+     *
+     * @override
+     * @protected
+     * @returns {boolean}
+     */
+    _isDisabled() {
+        return false;
+    }
+
+    /**
+     * Indicates whether this button is in toggled state or not.
+     *
+     * @override
+     * @protected
+     * @returns {boolean}
+     */
+    _isToggled() {
+        return this.props._isRecordingRunning;
+    }
+}
+
+/**
+ * Maps (parts of) the redux state to the associated props for the
+ * {@code RecordButton} component.
+ *
+ * @param {Object} state - The Redux state.
+ * @private
+ * @returns {{
+ *     _isRecordingRunning: boolean,
+ *     visible: boolean
+ * }}
+ */
+export function _mapStateToProps(state: Object): Object {
+    const isModerator = isLocalParticipantModerator(state);
+    const {
+        enableFeaturesBasedOnToken,
+        fileRecordingsEnabled
+    } = state['features/base/config'];
+    const { features = {} } = getLocalParticipant(state);
+
+    return {
+        _isRecordingRunning:
+            Boolean(getActiveSession(state, JitsiRecordingConstants.mode.FILE)),
+        visible: isModerator
```

Comment:
`visible` is an "own prop", not a prop designed to be passed as part of mapStateToProps. I think that we should use that logic in `isDiabled`.

Reply:
Can you please explain it a bit more

Reply:
`visible` is designed to pass it directly to the button component, not for being computed. Basically so we check if the button was not on the list in `interfaceConfig`.

Redux will compute the props passed to the component as follows: `Object.assign({}, ownProps, stateProps, dispatchProps` by default. So, if you compute `visible` yourself, then you'd override the user's desire, since when they specify it, it will be part of `ownProps`.

Reply:
I don't agree that the containing component should tell the record button when to be visible and when not. If it wants to override it, that's fine, and thanks for pointing it out, I'll add a check using ownProps, but generally the button should contain all feature-specific logic in itself.